### PR TITLE
WIP: add new backend api documentation

### PR DESCRIPTION
**Pull request by aurghs**
- add backend documentation
- rename ``store_spec`` in ``filename_or_obj``in  backend entrypoint method ``guess_can_open`` 

- [x] Related #4803

### CODE REVIEW

File: doc/internals.rst [hunk 1]
```diff
@@ -231,3 +231,101 @@ re-open it directly with Zarr:
...
+------------------------
+
+Adding a new backend for read support to Xarray is easy, and you don't need to integrate your code in Xarray.
+All you need to do is to:
+
+- Implement a function that returns an instance :py:class:``Dataset``
+
+- Create a `BackendEntrypoint`` instance with your function as input.
+
+- Declare such instance as external plugin in your setup.py.
```
Comment (Illviljan) [inline doc/internals.rst:246]: As someone who's been working on opening custom hdf formats in xarray, declaring an external plugin in setup.py doesn't seem _easy_ to me.

Why can't we simply be allowed to add functions to the engine parameter? Example:

```python
from custom_backend import engine

ds = xr.load_dataset(filename, engine=engine)
```

Comment (aurghs) [reply doc/internals.rst:246]: I see our point, and I think we can also add an option to open_dataset to take in input the customed backend class.

Comment (jhamman) [reply doc/internals.rst:246]: @Illviljan - this is an interesting idea but probably deserves a more detailed discussion. Can you (or @aurghs open a separate issue to discuss)?

Comment (Illviljan) [reply doc/internals.rst:246]: Done, see #4838.

File: doc/internals.rst [hunk 2]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+How to add a new backend
+------------------------------------
+
+Adding a new backend for read support to Xarray is easy, and does not require
+to integrate any code in Xarray; all you need to do is approaching the
+following steps:
+
+- Create a class that inherits from Xarray py:class:`~xarray.backend.commonBackendEntrypoint`
+- Implement the method ``open_dataset`` that returns an instance of :py:class:`~xarray.Dataset`
+- Declare such a class as an external plugin in your setup.py.
```
Comment (dcherian) [inline doc/internals.rst:245]: Can we link to the entrypoints documentation here.

```suggestion
- Declare such a class as an external plugin in your ``setup.py``.
```

EDIT: I see it's linked below. Won't hurt to add a link here also.

File: doc/internals.rst [hunk 3]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+
+These are detailed in the following.
+
+open_dataset
+++++++++++++
+
+Inputs
+^^^^^^
+
+The backend ``open_dataset`` method takes as input one argument
```
Comment (dcherian) [inline doc/internals.rst:262]: So these are two required arguments?

File: doc/internals.rst [hunk 4]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+(``filename``), and one keyword argument (``drop_variables``):
+
+- ``filename``: can be a string containing a relative path or an instance of ``pathlib.Path``.
+- ``drop_variables``: can be `None` or an iterable containing the variable names to be dropped when reading the data.
+
+If it makes sense for your backend, your ``open_dataset`` method should
+implement in its interface all the following boolean keyword arguments, called
+**decoders** which default to ``None``:
+
+- ``mask_and_scale=None``
```
Comment (dcherian) [inline doc/internals.rst:272]: and these are optional kwargs?

File: doc/internals.rst [hunk 5]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+- ``use_cftime=None``
+- ``concat_characters=None``
+- ``decode_coords=None``
+
+These keyword arguments are explicitly defined in Xarray
+:py:meth:`~xarray.open_dataset` signature.  Xarray will pass them to the
+backend only if the User sets a value different from ``None`` explicitly.
+Your backend can also take as input a set of backend-specific keyword
+arguments. All these keyword arguments can be passed to
+:py:meth:`~xarray.open_dataset` grouped either via the ``backend_kwarg``
```
Comment (dcherian) [inline doc/internals.rst:284]:
```suggestion
:py:meth:`~xarray.open_dataset` grouped either via the ``backend_kwargs``
```
i think it is "kwargs" not "kwarg"?

File: doc/internals.rst [hunk 6]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+parameter or explicitly using the syntax ``**kwargs``.
+
+Output
+^^^^^^
+The output of the backend `open_dataset` shall be an instance of
+Xarray py:class:`~xarray.Dataset` that implements the additional method ``close``,
```
Comment (dcherian) [inline doc/internals.rst:290]:
```suggestion
Xarray :py:class:`~xarray.Dataset` that implements the additional method ``close``,
```

File: doc/internals.rst [hunk 7]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+used by Xarray to ensure the related files are eventually closed.
+
+If you don't want to support the lazy loading, then the :py:class:`~xarray.Dataset`
+shall contain ``NumPy.arrays`` and your work is almost done.
```
Comment (dcherian) [inline doc/internals.rst:294]:
```suggestion
shall contain :py:class:`numpy.ndarray` and your work is almost done.
```

File: doc/internals.rst [hunk 8]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+explicitly, Xarray creates a list of them automatically by inspecting the
+backend signature.
+
+Xarray uses ``open_dataset_parameters`` only when it needs to select
+the **decoders** supported by the backend.
+
+If ``open_dataset_parameters`` is not defined, but ``**kwargs`` and ``*args`` have
+been passed to the signature, Xarray raises an error.
+On the other hand, if the backend provides the ``open_dataset_parameters``,
+then ``**kwargs`` and `*args`` can be used in the signature.
```
Comment (dcherian) [inline doc/internals.rst:309]:
```suggestion
then ``**kwargs`` and ``*args`` can be used in the signature.
```

File: doc/internals.rst [hunk 9]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+py:meth:`~xarray.backend.common BackendEntrypoint.guess_engine` that always returns ``False``.
+
+Backend ``guess_can_open`` takes as input the ``filename_or_obj`` parameter of
+Xarray :py:meth:`~xarray.open_dataset`, and returns a boolean.
+
+
+How to register a backend
++++++++++++++++++++++++++++
+
+Define in your setup.py (or setup.cfg) a new entrypoint with:
```
Comment (dcherian) [inline doc/internals.rst:329]:
```suggestion
Define in your ``setup.py`` (or ``setup.cfg``) a new entrypoint with:
```

File: doc/internals.rst [hunk 10]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+
+How to support Lazy Loading
++++++++++++++++++++++++++++
+If you want to make your backend effective with big datasets, then you should support
+the lazy loading.
+Basically, you shall replace the :py:class:`numpy.array` inside the variables with
+a custom class:
+
+.. ipython:: python
+    backend_array = YourBackendArray()
```
Comment (dcherian) [inline doc/internals.rst:346]: This needs to be "verbatim" otherwise sphinx tried to execute it, that's why the RTD build failed.

File: doc/internals.rst [hunk 11]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+    data = indexing.LazilyOuterIndexedArray(backend_array)
+    variable = Variable(..., data, ...)
+
+Where ``YourBackendArray``is a class that inherit from
```
Comment (dcherian) [inline doc/internals.rst:350]: Can we add an example ``YourBackendArray`` in the beginning that shows the signature for ``open_dataset``, the optional attributes, and the use of optional ``kwargs``? It would make this document easier to understand

File: doc/internals.rst [hunk 12]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+inside the variable’s encoding ``var.encoding["preferred_chunks"]``.
+The ``preferred_chunks`` may be useful to improve performances with lazy loading.
+``preferred_chunks`` shall be a  dictionary specifying chunk size per dimension
+like ``{“dim1”: 1000, “dim2”: 2000}``  or
+``{“dim1”: [1000, 100], “dim2”: [2000, 2000, 2000]]}``.
+
+The ``preferred_chunks`` is used by Xarray to define the chunk size in some
+special cases:
+
+- If ``chunks`` along a dimension is None or not defined
```
Comment (dcherian) [inline doc/internals.rst:375]:
```suggestion
- If ``chunks`` along a dimension is ``None`` or not defined
```

File: doc/internals.rst [hunk 13]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+- If ``chunks`` is “auto”
```
Comment (dcherian) [inline doc/internals.rst:376]:
```suggestion
- If ``chunks`` is ``“auto”``
```

File: doc/internals.rst [hunk 14]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+
+In the first case Xarray uses the chunks size specified in
+``preferred_chunks``.
+In the second case Xarray accommodates ideal chunk sizes, preserving if
+possible the "preferred_chunks". The ideal chunk size is computed using
+``dask.core.normalize function``, setting ``previus_chunks = preferred_chunks``.
```
Comment (dcherian) [inline doc/internals.rst:382]:
```suggestion
:py:func:`dask.core.normalize_chunks`, setting ``previus_chunks = preferred_chunks``.
```

File: doc/internals.rst [hunk 15]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+The decoders implement specific operations to transform data from on-disk
+representation to Xarray representation.
+
+A classic example is the “time” variable decoding operation. In NetCDF, the
+elements of the “time” variable are stored as integers, and the unit contains
+an origin (for example: "seconds since 1970-1-1"). In this case, Xarray
+transforms the pair integer-unit in a ``np.datetimes``.
+
+The standard decoders implemented in Xarray are:
+- strings.CharacterArrayCoder()
```
Comment (dcherian) [inline doc/internals.rst:396]: the list needs some formatting.

File: doc/internals.rst [hunk 16]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+- variables.CFMaskCoder()
+- variables.CFScaleOffsetCoder()
+- times.CFTimedeltaCoder()
+- times.CFDatetimeCoder()
+
+Some of the transformations can be common to more backends, so before
+implementing a new decoder, be sure Xarray does not already implement that one.
+
+The backends can reuse Xarray’s decoders, either instantiating the decoders
+directly or using the higher-level function
```
Comment (dcherian) [inline doc/internals.rst:408]: Here again an example showing the use of one built-in decoder function would be useful.

**Review by dcherian — COMMENTED** [review]
Thanks @aurghs 

I have mostly minor suggestions.

One high-level comment is that it would be nice to have an example implementation with pseudo-code. That would make it easier to understand.


``` python
class YourBackendArray(BackendEntrypoint):
    def open_dataset(...):
        ds = ...
 	    # example built-in decoder use
        return ds

    open_dataset_parameters = ...

    def guess_can_open(filename_obj):
        # do checks
        return True

```

File: doc/internals.rst [hunk 17]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+- name: the name to be passed to :py:meth:`~xarray.open_dataset`  as ``engine``.
+- object reference: the reference of the class that you have implemented.
+
+See https://packaging.python.org/specifications/entry-points/#data-model
+for more information
+
+How to support Lazy Loading
++++++++++++++++++++++++++++
+If you want to make your backend effective with big datasets, then you should support
+the lazy loading.
```
Comment (mathause) [inline doc/internals.rst:341]:
```suggestion
lazy loading.
```

File: doc/internals.rst [hunk 18]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
     zgroup = zarr.open("rasm.zarr")
     print(zgroup.tree())
     dict(zgroup["Tair"].attrs)
+
+
+How to add a new backend
+------------------------------------
+
+Adding a new backend for read support to Xarray is easy, and does not require
```
Comment (mathause) [inline doc/internals.rst:239]:
```suggestion
Adding a new backend for read support to Xarray does not require
```
(easy is relative so I'd leave this out)

File: doc/internals.rst [hunk 19]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+to integrate any code in Xarray; all you need to do is approaching the
+following steps:
```
Comment (mathause) [inline doc/internals.rst:241]:
```suggestion
to integrate any code in Xarray; all you need to do is follow these steps:
```

File: doc/internals.rst (hunk 2 above)
Comment (mathause) [inline doc/internals.rst:245]:
```suggestion
- Declare this class as an external plugin in your setup.py.
```

File: doc/internals.rst [hunk 20]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+
+- Create a class that inherits from Xarray py:class:`~xarray.backend.commonBackendEntrypoint`
```
Comment (mathause) [inline doc/internals.rst:243]:
```suggestion
- Create a class that inherits from Xarray py:class:`~xarray.backends.common.BackendEntrypoint`
```

File: doc/internals.rst (hunk 10 above)
Comment (keewis) [reply doc/internals.rst:346]: actually, the current failure is due to a missing newline between the rst directive and the code:
```suggestion

    backend_array = YourBackendArray()
```
once that is fixed it will indeed raise a `NameError`

File: doc/internals.rst [hunk 21]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+
+However, this practice is discouraged unless there is a good reasons for using
+`**kwargs` or `*args`.
+
+guess_can_open
+++++++++++++++
+``guess_can_open`` is used to identify the proper engine to open your data
+file automatically in case the engine is not specified explicitly. If you are
+not interested in supporting this feature, you can skip this step since
+py:class:`~xarray.backend.common.BackendEntrypoint` already provides a default
```
Comment (mathause) [inline doc/internals.rst:319]:
```suggestion
py:class:`~xarray.backends.common.BackendEntrypoint` already provides a default
```
I think that's `backends`

**Review by mathause — COMMENTED** [review]
I also think that an example would go a long way. It could also be a good idea to point to the simplest of the implemented backends.

I tried to look at some of the backends that are now available, e.g. [pydap](https://github.com/pydata/xarray/blob/5735e163bea43ec9bc3c2e640fbf25a1d4a9d0c0/xarray/backends/pydap_.py) and one thing that's missing from the description here is the `DataStore`.

Not for here but I did also not find much on what a `LazilyOuterIndexedArray` is etc...

File: doc/internals.rst [hunk 22]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+
+How to register a backend
++++++++++++++++++++++++++++
+
+Define in your setup.py (or setup.cfg) a new entrypoint with:
+
+- group: ``xarray.backend``
+- name: the name to be passed to :py:meth:`~xarray.open_dataset`  as ``engine``.
+- object reference: the reference of the class that you have implemented.
+
```
Comment (keewis) [inline doc/internals.rst:334]: it would probably help to have a example here, too

**Review by keewis — COMMENTED** [review]
I agree with the need for examples, but overall I think this is pretty well structured. In case it is of use, I think this pretty helpful: https://documentation.divio.com/how-to-guides/

I think it would be useful to have a comparison between the old and new API, but I'm not sure if the old API was part of the public API (if not this is not necessary).

`BackendArray`, `CachingFileManager`, `LazilyOuterIndexedArray`, and `LazilyVectorizedIndexedArray` are not yet public API (I think?), so if we want to change that I think we should make sure their purpose is properly documented (it looks like that's what you were going to do, though).

This makes `internals` a fairly long document, so we should think about splitting this into different pages (not in this PR). I think @andersy005 started something like that in #4835.

File: doc/internals.rst (hunk 10 above)
Comment (keewis) [reply doc/internals.rst:346]: it seems this exposed a bug in `blackdoc` (which is why the linting CI failed). I'll look it.

File: doc/internals.rst (hunk 17 above)
Comment (aurghs) [reply doc/internals.rst:341]: > Not for here but I did also not find much on what a `LazilyOuterIndexedArray` is etc...

I am still working on this part. I tried to gather some information, but couldn't find much around, unfortunately. I'm still putting the pieces together to get a better idea of how it works.

File: doc/internals.rst (hunk 3 above)
Comment (aurghs) [reply doc/internals.rst:261]: Yes

File: doc/internals.rst [hunk 23]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+Xarray uses ``open_dataset_parameters`` only when it needs to select
+the **decoders** supported by the backend.
+
+If ``open_dataset_parameters`` is not defined, but ``**kwargs`` and ``*args`` have
+been passed to the signature, Xarray raises an error.
+On the other hand, if the backend provides the ``open_dataset_parameters``,
+then ``**kwargs`` and `*args`` can be used in the signature.
+
+However, this practice is discouraged unless there is a good reasons for using
+`**kwargs` or `*args`.
```
[the remark this answers is not in the document]
Comment (aurghs) [reply doc/internals.rst]: > I tried to look at some of the backends that are now available, e.g. [pydap](https://github.com/pydata/xarray/blob/5735e163bea43ec9bc3c2e640fbf25a1d4a9d0c0/xarray/backends/pydap_.py) and one thing that's missing from the description here is the `DataStore`.

`DataStore` is not part of the interface. Currently, it no more needed. The old backends still have the `DataStore` but it is an internal implementation detail.

File: doc/internals.rst (hunk 4 above)
Comment (aurghs) [reply doc/internals.rst:271]: yes

File: doc/internals.rst (hunk 22 above)
Comment (aurghs) [reply doc/internals.rst:449]: added

File: doc/internals.rst (hunk 16 above)
Comment (aurghs) [reply doc/internals.rst]: done

File: doc/internals.rst [hunk 24]
```diff
@@ -231,3 +231,410 @@ re-open it directly with Zarr:
...
+    >>> backend_array._raw_indexing_method(())
+    array([[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11]])
+
+    # shall support integers
+    >> backend_array._raw_indexing_method(1, 1)
+    5
+
+    # shall support slices
+    >>> backend_array._raw_indexing_method(slice(0, 3), slice(2, 4))
+    array([[2, 3], [6, 7], [10, 11]])
```
Comment (keewis) [inline doc/internals.rst:591]: while `sphinx` supports this, `blackdoc` was not written with something like this in mind (and adding this will probably take some time). Since we normally use `ipython` instead, how about converting this to verbatim `ipython`? For example, this would become:
```suggestion
.. ipython::
    :verbatim:
    
    In [1]: # () shall return the full array
       ...: backend_array._raw_indexing_method(())
    Out[1]: array([[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11]])
    
    In [2]: # shall support integers
       ...: backend_array._raw_indexing_method(1, 1)
    Out[2]: 5
    
    In [3]: # shall support slices
       ...: backend_array._raw_indexing_method(slice(0, 3), slice(2, 4))
    Out[3]: array([[2, 3], [6, 7], [10, 11]])
```

Comment (aurghs) [reply doc/internals.rst:591]: thank you very much for your suggestions!
However, it is not worth reviewing the new documentation now, sincee the interface for the lazy loading part will probably be changed.

Comment (keewis) [reply doc/internals.rst:591]: sure. This is more a hint of how to get the linting CI to pass.

**Comment by jhamman** [comment]
@jp-dark - Given your recent experience developing a backend using this new API, it would be great if you could provide some comments on this PR.

**Comment by kmuehlbauer** [comment]
@jhamman As I'm currently implementing xarray backends for reading weather radar data, I've already closely followed the adventures of @aurghs and @alexamici.

To the current state of the backend documentation I can say that it is very well written and understandable (also for non native speakers) and I could easily follow the train of thought.

From my perspective: I would not try to show extensive examples in the beginning, but would extend/rearrange documentation after first users issued problems/asked questions. So that only things get special documentation which are not immediately clear from users perspective.

This is a really remarkable extension which will boost usage of xarray. Thanks!

File: doc/internals.rst [hunk 25]
```diff
@@ -231,3 +231,448 @@ re-open it directly with Zarr:
...
+
+**OUTER_1VECTOR**
+
+The ``OUTER_1VECTOR`` indexing shall supports number, slices and at least one
+list. The behaviour with the list shall be the same of ``OUTER`` indexing.
+
+If you support more complex indexing as `explicit indexing` or
+`numpy indexing`, you can have a look to the implemetation of Zarr backend and Scipy backend,
+currently available in :py:mod:`~xarray.backends` module.
+
```
Comment (jp-dark) [inline doc/internals.rst:652]: It may be useful to add a discussion here on how a raw key is transformed to an `ExplicitIndexer` when querying an xarray `DataArray` with a third-party backend. This would be especially useful for picking good test cases for querying `DataArray`s and debugging problems in edge-cases.

File: doc/internals.rst [hunk 26]
```diff
@@ -231,3 +231,448 @@ re-open it directly with Zarr:
...
+        def __getitem__(self, key):
+            return indexing.explicit_indexing_adapter(
+                key,
+                self.shape,
+                indexing.IndexingSupport.BASIC,
+                self._raw_indexing_method,
+            )
+
+        def _raw_indexing_method(self, key):
```
Comment (shoyer) [inline doc/internals.rst:575]:
```suggestion
        def _raw_indexing_method(self, key: tuple) -> np.typing.ArrayLike:
```

(could also note that in this case, `key` could be `Tuple[Union[int, slice], ...]` because it only implements "basic" indexing)

File: doc/internals.rst [hunk 27]
```diff
@@ -231,3 +231,448 @@ re-open it directly with Zarr:
...
+    Out[1]: array([[0, 1, 2], [4, 5, 6]])
+
+    # shall support integers
+    In [2]: backend_array._raw_indexing_method(1, 1)
+    Out[2]: 5
+
+
+**OUTER_1VECTOR**
+
+The ``OUTER_1VECTOR`` indexing shall supports number, slices and at least one
```
Comment (shoyer) [inline doc/internals.rst:646]: This should be "at most" rather than "at least"

File: doc/internals.rst [hunk 28]
```diff
@@ -231,3 +231,448 @@ re-open it directly with Zarr:
...
+.. code-block:: python
+
+    backend_array = MyBackendArray()
+    data = indexing.LazilyOuterIndexedArray(backend_array)
+    var = xr.Variable(dims, data, attrs=attrs, encoding=encoding)
+
+Where:
+
+- :py:class:`~xarray.core.indexing.LazilyOuterIndexedArray` is a class
+  provided by Xarray that manages the lazy loading.
```
Comment (shoyer) [inline doc/internals.rst:524]: To help make things easier to understand, could we rename this class from `LazilyOuterIndexedArray` to the more generic name `LazilyIndexedArray`? (It currently works in all cases, not just outer indexing)

File: doc/internals.rst [hunk 29]
```diff
@@ -231,3 +231,448 @@ re-open it directly with Zarr:
...
+            shape,
+            dtype,
+            lock,
+            # other backend specific keyword arguments
+        ):
+            self.shape = shape
+            self.dtype = lock
+            self.lock = dtype
+
+        def __getitem__(self, key):
```
Comment (shoyer) [inline doc/internals.rst:567]: I would suggest adding a type annotation:
```suggestion
        def __getitem__(self, key: xarray.core.indexing.ExplicitIndexer) -> np.typing.ArrayLike:
```

File: doc/internals.rst (hunk 27 above)
Comment (aurghs) [reply doc/internals.rst:646]: definitely! sorry for the over oversight

File: doc/internals.rst (hunk 28 above)
Comment (aurghs) [reply doc/internals.rst]: done

File: doc/internals.rst (hunk 27 above)
Comment (aurghs) [reply doc/internals.rst]: fixed

File: doc/internals.rst [hunk 30]
```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
...
+- Create a class that inherits from Xarray py:class:`~xarray.backend.commonBackendEntrypoint`
+- Implement the method ``open_dataset`` that returns an instance of :py:class:`~xarray.Dataset`
+- Declare such a class as an external plugin in your setup.py.
+
+Your ``BackendEntrypoint`` sub-class is the primary interface with Xarray, and
+it should implement the following attributes and functions:
+
+- ``open_dataset`` (mandatory)
+- [``open_dataset_parameters``] (optional)
+- [``guess_can_open``] (optional)
```
[the remark this answers is not in the document]
Comment (aurghs) [reply doc/internals.rst]: done

File: doc/internals.rst (hunk 25 above)
Comment (aurghs) [reply doc/internals.rst:654]: I can try to add something about it.

File: doc/internals.rst [hunk 31]
```diff
@@ -231,3 +231,450 @@ re-open it directly with Zarr:
...
+.. code-block:: cfg
+
+    [options.entry_points]
+    xarray.backends =
+        engine_name = your_package.your_module:your_backendentrypoint
+
+
+See https://packaging.python.org/specifications/entry-points/#data-model
+for more information
+
```
Comment (jp-dark) [inline doc/internals.rst:504]:
```suggestion

If you are using [Poetry](https://python-poetry.org/) for your build system, you can accomplish the same thing using "plugins". In this case you would need to add the following to your ``pyproject.toml`` file:

.. code-block:: toml

    [tool.poetry.plugins."xarray_backends"]
    "engine_name" = "your_package.your_module:YourBackendEntryClass"
    
See https://python-poetry.org/docs/pyproject/#plugins for more information of Poetry plugins.
```

File: doc/internals.rst [hunk 32]
```diff
@@ -231,3 +231,450 @@ re-open it directly with Zarr:
...
+        },
+    )
+
+in ``setup.cfg``:
+
+.. code-block:: cfg
+
+    [options.entry_points]
+    xarray.backends =
+        engine_name = your_package.your_module:your_backendentrypoint
```
Comment (jp-dark) [inline doc/internals.rst:499]:
```suggestion
                "engine_name=your_package.your_module:YourBackendEntryClass"
            ],
        },
    )

in ``setup.cfg``:

.. code-block:: cfg

    [options.entry_points]
    xarray.backends =
        engine_name = your_package.your_module:YourBackendEntryClass
```

The above suggestion is a small naming change to emphasize the entrypoint object is a class.

Comment (aurghs) [reply doc/internals.rst]: Thank you! You are right!

File: doc/internals.rst (hunk 31 above)
Comment (aurghs) [reply doc/internals.rst:504]: Is poetry widely used? I have no idea of this, so I do not know whether to add it. @alexamici what do you suggest?

File: doc/internals.rst (hunk 25 above)
Comment (aurghs) [reply doc/internals.rst:654]: I had a look at what to add here. I started writing some more details, but I think It would become a little bit too long. 
Today we should merge. So I suggest to merge it as it is, and I'll add later a dedicated section.
